### DISCUSSION TITLE
"Safe" and "unsafe" fraction when shape-casting

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#450] [Q&A] [answered]
I've run into an interesting problem with implementing shape-casting in my Godot implementation that I'm curious to hear your take on. Apologies in advance for the lengthy question.

The way shape-casts are designed in Godot, for reasons that are unknown to me, essentially makes them a two-stage process if you want any sort of information out of them that isn't just the fraction. The scripting interface for [`cast_motion`](https://docs.godotengine.org/en/latest/classes/class_physicsdirectspacestate3d.html#class-physicsdirectspacestate3d-method-cast-motion) only returns two floats, which represent what they refer to as the "safe" and "unsafe" fractions of the cast motion. These two floats are explained as "the maximum fraction of the motion that can be made without a collision" and "the minimum fraction of the distance that must be moved for a collision", respectively.

This means that to get something like the collision point or collision normal you have to move your query transform along the cast motion by the unsafe fraction and perform something like `JPH::NarrowPhaseQuery::CollideShape` using what they call [`get_rest_info`](https://docs.godotengine.org/en/latest/classes/class_physicsdirectspacestate3d.html#class-physicsdirectspacestate3d-method-get-rest-info).

When looking at [their Bullet implementation](https://github.com/godotengine/godot/blob/e90ac4b0e7636d6ac7d1f658fc7984f70d6fb7f3/modules/bullet/space_bullet.cpp#L195-L196) of `cast_motion` it seems like Bullet's `convexSweepTest` returns what it refers to as `m_closestHitFraction`, which is passed on unmodified as the unsafe fraction, and then they just nudge that unsafe fraction by 0.01 units along the cast motion and pass that result on as the safe fraction. This strikes me as a bit error-prone, but I guess it should work in most cases. Either way, the unsafe fraction is the important part I guess.

I tried taking the same approach as their Bullet implementation, but found that the result of performing a `CollideShape` query at the fraction returned from `CastShape` (both coupled with a "closest hit" collector) was somewhat erratic depending on the shape being hit, which I took to mean that I can't reliably use it as an unsafe nor safe fraction. One workaround for this is to simply interpret the fraction returned from `CastShape` as being perfectly at the edge of the safe and unsafe fraction and just nudge it by some arbitrary amount (like 0.005) in _both_ directions to get the two fractions.

What's your take on this? Is there perhaps some way to retrieve both of these fractions from Jolt without doing this arbitrary nudging? Is it an impossible ask due to floating-point precision? Could I be doing something wrong with the either of the two queries?

[Here](https://github.com/godotengine/godot/blob/3863199ab940272f6844ff30910ec7a520e47f41/servers/physics_3d/godot_space_3d.cpp#L259-L376) is their implementation of `cast_motion` in Godot Physics, in case you're curious.

[Here](https://github.com/mihe/JoltPhysics/blob/e86f9f67385833584c8749b3ed49cbedc381a86d/Samples/Tests/General/CastShapeTest.cpp) you'll find a simple test I created to show the erratic behavior that I described above, which is what you can see in the clip below. Increasing the height of the floor box seems to fix the erratic behavior.

https://user-images.githubusercontent.com/4884246/221355701-2120cc5d-9a4c-45b1-827b-735bea834c21.mp4

**Comment by jrouwe** [comment, marked as answer]
Hello,

If I look at the godot implementation then they implemented cast_motion as a binary search, so they automatically get a low and a high value where high is colliding and low is not. They only seem to take 8 steps, so there's is a fixed precision of 2^-8 * length_of_cast, which imo could be pretty inaccurate. 

Jolt (and I think also Bullet) use a conservative advancement method which iteratively takes the maximum known safe step towards the two objects colliding and it continues doing so until the distance between the two objects is within tolerance (ShapeCastSettings::mCollisionTolerance). Since it's conservative, it will try to err on the 'non-touching' side, so if you then take that fraction and do a CollideShape there's indeed a chance that it does not find a collision. Even if it was possible to specify a tolerance of 0, the algorithms used in CastShape and CollideShape are slightly different, so I don't think you can ever guarantee that the returned fraction will result in a collision or not result in a collision.

The question then is indeed: What will work? I think you can use mCollisionTolerance to nudge the point forward / backward to get the guaranteed non-colliding / colliding fractions although you may actually need a little bit more than mCollisionTolerance because working with floats gives round offs everywhere, so even the tolerance check is not 100% accurate. In any case I think it needs 'tuning' of an epsilon.

Alternatively you could implement cast_motion in the exact same way as Godot by doing a CastAABox on the broad phase and then for every body found do a binary search using CollideShape to find the two fractions, but it's likely that that will be less performant than the current implementation.

**Reply by mihe** [reply]
> They only seem to take 8 steps, so there's is a fixed precision of 2^-8 * length_of_cast, which imo could be pretty inaccurate.

Interesting. That would actually explain why the output from their shape-casting is so jittery when doing a similar oscillating motion as what I showed in the clip above. I thought it was a bug, but I guess it's a limitation in their approach.

> I think you can use mCollisionTolerance to nudge the point forward / backward to get the guaranteed non-colliding / colliding fractions

When you say "use mCollisionTolerance to nudge the point forward / backward", I assume you mean to use that as the distance by which to nudge the final fraction in either direction? That makes a lot more sense than using some arbitrary fixed distance I guess. But yes, some epsilon probably needs to be applied there as well.

> Alternatively you could implement cast_motion in the exact same way as Godot

That's interesting. I hadn't even considered that to be an option, but I guess it really isn't much more than just that broad phase check and a bunch of `CollideShape` calls. Although I'm not sure I want to end up with the same accuracy problems as they have. I assume increasing the number of steps isn't gonna make it all that much better in terms of precision/stability either, not to mention performance.

I'll probably end up going for the tolerance nudging option. On top of that I think I'll go against what the Godot documentation says about `get_rest_info` where they state that "This method does not take into account the `motion` property". It just so happens that they use the same struct of parameters for that method as they do for `cast_motion`, which means I can conditionally do a `CollideShape` or `CastShape` depending on if `motion` is zero. That would let people skip the whole two-stage process if they really want/need to.

Thank you!

**Reply by jrouwe** [reply]
> I assume you mean to use that as the distance by which to nudge the final fraction in either direction

Correct.

> I assume increasing the number of steps isn't gonna make it all that much better in terms of precision/stability either, not to mention performance.

More steps is definitively bad for performance, but with each step you make the result a factor 2 more accurate so it will become more accurate very quickly.

**Reply by mihe** [reply]
> I think you can use mCollisionTolerance to nudge the point forward / backward to get the guaranteed non-colliding / colliding fractions although you may actually need a little bit more than mCollisionTolerance because working with floats gives round offs everywhere, so even the tolerance check is not 100% accurate. In any case I think it needs 'tuning' of an epsilon.

I'm noticing that this approach only seems to hold up when dealing with a convex radius of 0.

When using any other convex radius for the two shapes then I'm more or less forced to use `mUseShrunkenShapeAndConvexRadius` when casting, which I guess makes sense, but I then also have to nudge the resulting fraction not only by the collision tolerance but also by the convex radius of the shape that's being cast, in order to produce the unsafe/touching fraction.

This seems to indicate that `CastShape` produces a fraction that's as if there was no convex radius on the shape being cast, which I then need to compensate for.

Am I understanding this correctly, and is this the intended behavior?

**Reply by jrouwe** [reply]
No, the convex radius should be taken into account while sweeping. I just ran some tests using the probe in the SamplesApp and as far as I can see it is working.

I have to say that the convex radius in Jolt may work slightly different from other engines, as when you create a shape it does not expand the shape with the convex radius but instead shrinks the shape and then expands it again. For a box vs box sweep this means:

![image](https://user-images.githubusercontent.com/1621693/226170900-dd298169-4dd0-4b9a-9ca1-70d722bd40bd.png)

On the left is the default behavior (with exaggerated convex radius). The value of the convex radius is irrelevant in this case as the cast shape function just uses the original shape to do the casting. On the right is when you use the mUseShrunkenShapeAndConvexRadius option. In this case both shapes are shrunken by their convex radius and then expanded again, meaning that you're actually colliding against the inner green/blue shapes with rounded corners.

During CollideShape tests the convex radius matters more as by default we will use the inner green/blue shapes for collision detection and only if the collision detection algorithm finds that there is an overlap, the colliding points + penetration depth will be calculated using the real shapes (so there is the possibility there that you find that two boxes are not overlapping while they actually are).

What kind of shapes are you colliding? E.g. it is quite difficult to properly shrink a convex hull shape, so if you first do a cast shape and then a collide shape you may find differences in collision between the two.

**Reply by mihe** [reply]
It is indeed convex hulls that are creating problems. Though it seems like my theory about the convex radius being needed for this "nudge" was mostly erroneously reinforced by a particular test case that just happened to (kinda) work with certain values/shapes. As soon as I change the convex radius for the convex hull, or even remove the convex radius altogether, then things fall apart pretty quickly. So it seems like this nudge approach never really worked well when using convex hulls as the shape that's being cast.

In case you're curious, I've set up a branch [here](https://github.com/mihe/JoltPhysics/blob/97ee808c5d04024740eb55b1298ae6cf4c290202/Samples/Tests/General/CastShapeTest.cpp) with another `Samples` test that shows where these nudged fractions fall apart. The test consists of three shapes (sphere, box, hull) being cast down at varying heights onto the same three shapes, but scaled up. If either of the two nudged fractions fail their respective `CollideShape` call then a red line is drawn at that position.

When using a convex radius of 0.05 on both the box and hull shape, and nudging based on that amount, it looks like this:

https://user-images.githubusercontent.com/4884246/226210641-56058acc-4731-4b2e-960d-a7c30275c06a.mp4

When instead setting the convex radius to 0.0, or even something like 0.01, it looks like this:

https://user-images.githubusercontent.com/4884246/226210932-57c69e76-8bba-4eae-95a3-5fdce37ce8f0.mp4

Do you see any way to wrangle convex hulls into working reliably with this method, or would you say that I'm better off attempting some kind of `CollideShape` search, similar to what Godot themselves were doing, considering these differences in behavior between `CastShape` and `CollideShape`?

**Reply by mihe** [reply]
I ended up making an attempt at porting the binary search from the Godot implementation, which at first glance seems to work about as well as theirs, although I haven't looked at performance at all yet.

The only real stumbling block turned out to be creating the equivalent of their [`GodotMotionShape3D`](https://github.com/godotengine/godot/blob/0291fcd7b66bcb315a49c44de8031e5596de4216/servers/physics_3d/godot_shape_3d.h#L474-L511), which I guess is what extends the support by the motion vector, but I managed to make use of the convex sub shape types from #468 to create a sort of convex decorator shape that only implemented the methods necessary for `sCollideShapeVsShape`, which seems to be `GetLocalBounds` and `GetSupportFunction`, and which also contained a dedicated `SupportBuffer` for its inner shape.

As I understand it, using these convex sub shape types means that specialized collision functions like `sCollideSphereVsHeightField` won't be used, but since I'm effectively stretching the shapes anyway then I assume that's just how it has to be.

If any of that sounds like blasphemy then I'd love to know. :)

Assuming I don't run into any problematic discrepancies between this and the Godot implementation then I'll probably end up using this binary search stuff just to satisfy this particular interface. I'll look into exposing a shape-cast method that better matches Jolt's `CastShape` once I start adding features outside of what Godot expects, later down the road.

**Reply by jrouwe** [reply]
Sorry, forgot to reply on this one yesterday.

> If any of that sounds like blasphemy then I'd love to know. :)

No this is certainly not blasphemy, I looked at the code and it looks good to me (would probably have done the same).

**Reply by mihe** [reply]
Glad to hear it! Thank you for taking a look. Much appreciated.
